src/types.rs

```diff
@@ -44,8 +44,23 @@ pub enum Error {
     #[error("document task error: {0}")]
     DocumentTask(String),
 
-    #[error("query error: {0}")]
+    #[error("query: generic error: {0}")]
     Query(String),
+
+    #[error("dns: generic error: {0}")]
```

Comment:
I like the way we can simply create errors and return them in Result<> , but should we add them all inside a generic enum like this? Wouldn't we want the errors to be per component? (`src/dns/errors.rs`, `src/html5/errors.rs` etc)

Reply:
I think it'd make sense to make them component specific now that the list is growing (and will grow more in the future.) I remember this was added maybe a month or two back when there were only 2-3 error types, but now I'd be in favour to move groups of them into their own components

Reply:
I'm guessing the rule of thumb on when to create a new error type and when to stick with an existing type is when you want to match against a specific kind of error and not other errors.  So the number of error types becomes the number of specific kinds of error you need to match on in high-level application code.

The matching in error handling code is what is important here.  I can imagine needing a total of three kinds of error in all, or 20, depending on what kind of error handling is needed.